Projects/18_Clustering/Predict_Sales.py

```diff
@@ -0,0 +1,36 @@
+# import data for offers
+import pandas as pd
+df_offers = pd.read_excel("./WineKMC.xlsx", sheetname=0)
+df_offers.columns = ["offer_id", "campaign", "varietal", "min_qty", "discount", "origin", "past_peak"]
+
+# import data for transactions
+df_transactions = pd.read_excel("./WineKMC.xlsx", sheetname=1)
+df_transactions.columns = ["customer_name", "offer_id"]
+df_transactions['n'] = 1
+
+# estimate total sales per offer_id
+df_total_sales = pd.merge(df_offers['offer_id'].to_frame(), df_transactions)
+df_total_sales = df_total_sales.groupby(['offer_id']).sum()
+df_total_sales.columns = ['Total_Sales']
+df_total_sales = df_total_sales.reset_index()
+
+# create final DF
+df_offers_and_sales = pd.merge(df_offers,df_total_sales)
+
+# encode categorical predictors
+from sklearn import preprocessing
+le = preprocessing.LabelEncoder()
```

Comment:
Using `LabelEncoder` for categorical data is problematic unless your data is ordinal which means that your feature has some inherent ordering. For example encoding good, ok, bad as 2, 1, 0 is OK because the scale corresponds to some ordering in the concept. However encoding John, Jane, Julie as 0, 1, 2 does not make sense since the data is nominal. This may not be an issue for modeling sometimes but it can lead to erroneous models and makes interpretation faulty. Use One Hot Encoding to avoid this problem.

I've written about this [here](https://medium.com/@amirziai/a-flask-api-for-serving-scikit-learn-models-c8bcdaa41daa#.j0j35dryb) if you're interested in learning more. Here' another [good resource](https://www.quora.com/In-scikit-learn-what-is-the-best-way-to-handle-categorical-features-of-high-cardinality-Using-one-hot-encoder-seems-to-blow-up-my-feature-space-Does-assigning-a-number-to-each-cardinality-work-better-or-do-we-run-into-the-risk-of-the-ML-algorithm-assuming-some-ordering-in-these-numeric-labels). For larger datasets OHE is computationally prohibitive and you may want to use the [hashing trick](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.FeatureHasher.html).